### PR TITLE
Gkps state prep gkp complexcat fock

### PR DESCRIPTION
**Pull request by elib20**
State prep with gkps, complex cat representation, and fock

### CODE REVIEW

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 1]
```diff
@@ -187,8 +187,10 @@ def init_circuit(self, prog, **kwargs):
         self.circuit = BosonicModes()
         init_weights, init_means, init_covs = [[0] * nmodes for i in range(3)]
 
-        vac_means = np.zeros(2, dtype=complex).tolist()
-        vac_covs = np.identity(2, dtype=complex).tolist()
+        vac_means = np.zeros((1, 2), dtype=complex)  # .tolist()
+        vac_covs = np.array(
+            [[[0.5 * self.circuit.hbar, 0], [0, 0.5 * self.circuit.hbar]]]
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:192]: maybe a bit clearer/shorter to write `np.array([0.5*self.circuit.hbar*np.identity(2)])` ?

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 2]
```diff
@@ -187,8 +187,10 @@ def init_circuit(self, prog, **kwargs):
...
+        )  # .tolist()
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:193]: This comment could be removed?

Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:193]:
```suggestion
        vac_covs = np.array([0.5*self.circuit.hbar*np.identity(2)])  # .tolist()
```

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 3]
```diff
@@ -246,7 +248,7 @@ def init_circuit(self, prog, **kwargs):
         # Tensor product of the weights.
         weights = kron_list(init_weights)
         # De-nest the means iterator.
-        means = np.array([[a for b in tup for a in b] for tup in mean_combs])
+        means = np.array([[a for b in tup for a in b] for tup in mean_combs], dtype=complex)
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:251]: itertools for the win!

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 4]
```diff
@@ -277,39 +279,187 @@ def reset(self, pure=True, **kwargs):
...
+    def prepare_cat(self, alpha, phi, desc="complex"):
         """ Prepares the arrays of weights, means and covs for a cat state"""
-        return
+
+        # case alpha = 0 -> prepare vacuum
+        if np.isclose(np.absolute(alpha), 0):
+            return (
+                np.array([[1]], dtype=complex),
+                np.array([[0, 0]], dtype=complex),
+                np.array([[[0.5 * self.circuit.hbar, 0], [0, 0.5 * self.circuit.hbar]]]),
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:313]: same here: maybe better to write `np.array([0.5*self.circuit.hbar*np.identity(2)])`

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 5]
```diff
@@ -277,39 +279,187 @@ def reset(self, pure=True, **kwargs):
...
+            elif desc == "complex":
+                raise ValueError("The complex description of GKP is not implemented")
+        else:
+            raise ValueError("Only square GKP are implemented for now")
 
     def prepare_fock(self, n, r=0.0001):
         """ Prepares the arrays of weights, means and covs of a Fock state"""
-        return
+        if 1 / r ** 2 < n:
+            raise ValueError("The parameter r**2={} is larger than n={}".format(r ** 2, n))
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:436]:
```suggestion
            raise ValueError("The parameter 1 / r ** 2={} is smaller than n={}".format(1 / r ** 2, n))
```

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 6]
```diff
@@ -401,20 +551,20 @@ def measure_heterodyne(self, mode, shots=1, select=None):
...
+        if np.isinstance(modes, int):
             modes = [modes]
 
-        # make sure number of modes matches shape of r and V
+        # make sure number of modes matches np.shape of r and V
         N = len(modes)
         if len(r) != 2 * N:
             raise ValueError("Length of means vector must be twice the number of modes.")
-        if V.shape != (2 * N, 2 * N):
+        if V.np.shape != (2 * N, 2 * N):
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:561]: I am not sure one can do `array.np.shape` ?

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 7]
```diff
@@ -401,20 +551,20 @@ def measure_heterodyne(self, mode, shots=1, select=None):
...
             raise ValueError(
-                "Shape of covariance matrix must be [2N, 2N], where N is the number of modes."
+                "np.shape of covariance matrix must be [2N, 2N], where N is the number of modes."
             )
 
         # convert xp-ordering to symmetric ordering
-        means = np.vstack([r[:N], r[N:]]).reshape(-1, order="F")
+        means = np.vstack([r[:N], r[N:]]).np.reshape(-1, order="F")
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:567]: similarly here with `reshape`

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 8]
```diff
@@ -452,7 +602,7 @@ def measure_fock(self, modes, shots=1, select=None, **kwargs):
         reduced_mean = mean[modes_idxs]
 
         # check we are sampling from a gaussian state with zero mean
-        if np.allclose(mu, np.zeros_like(mu)):
+        if np.allclose(mu, np.np.zeros_like(mu)):
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:605]: looks like a replacement went wrong!

**Review by nquesada — COMMENTED** [review]
Seems like a replacement in the source file went wrong. There are a bunch of places where there are `np.`something that should not be there.

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 9]
```diff
@@ -277,39 +279,187 @@ def reset(self, pure=True, **kwargs):
...
         """ Prepares the arrays of weights, means and covs for a cat state"""
-        return
+
+        # case alpha = 0 -> prepare vacuum
+        if np.isclose(np.absolute(alpha), 0):
+            return (
+                np.array([[1]], dtype=complex),
+                np.array([[0, 0]], dtype=complex),
+                np.array([[[0.5 * self.circuit.hbar, 0], [0, 0.5 * self.circuit.hbar]]]),
+            )
```
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:314]:
```suggestion
                np.array([0.5 * self.circuit.hbar * np.identity(2)])
```

File: strawberryfields/backends/bosonicbackend/backend.py (hunk 6 above)
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:561]:
```suggestion
        if V.shape != (2 * N, 2 * N):
```

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 10]
```diff
@@ -401,20 +551,20 @@ def measure_heterodyne(self, mode, shots=1, select=None):
...
-        # make sure number of modes matches shape of r and V
+        # make sure number of modes matches np.shape of r and V
         N = len(modes)
         if len(r) != 2 * N:
             raise ValueError("Length of means vector must be twice the number of modes.")
-        if V.shape != (2 * N, 2 * N):
+        if V.np.shape != (2 * N, 2 * N):
             raise ValueError(
-                "Shape of covariance matrix must be [2N, 2N], where N is the number of modes."
+                "np.shape of covariance matrix must be [2N, 2N], where N is the number of modes."
```
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:563]:
```suggestion
                "Shape of covariance matrix must be [2N, 2N], where N is the number of modes."
```

File: strawberryfields/backends/bosonicbackend/backend.py (hunk 7 above)
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:567]:
```suggestion
        means = np.vstack([r[:N], r[N:]]).reshape(-1, order="F")
```

File: strawberryfields/backends/bosonicbackend/backend.py (hunk 8 above)
Comment (elib20) [reply strawberryfields/backends/bosonicbackend/backend.py:605]:
```suggestion
        if np.allclose(mu, np.zeros_like(mu)):
```

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 11]
```diff
@@ -473,7 +623,7 @@ def measure_threshold(self, modes, shots=1, select=None, **kwargs):
         mu = self.circuit.mean
         cov = self.circuit.scovmatxp()
         # check we are sampling from a gaussian state with zero mean
-        if not np.allclose(mu, np.zeros_like(mu)):
+        if not np.allclose(mu, np.np.zeros_like(mu)):
```
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:626]:
```suggestion
        if not np.allclose(mu, np.zeros_like(mu)):
```

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 12]
```diff
@@ -310,8 +308,7 @@ def prepare_cat(self, alpha, phi, desc="complex"):
             return (
                 np.array([[1]], dtype=complex),
                 np.array([[0, 0]], dtype=complex),
-                np.array([[[0.5 * self.circuit.hbar, 0], [0, 0.5 * self.circuit.hbar]]]),
-            )
+                np.array([0.5 * self.circuit.hbar * np.identity(2)])
```
Comment (elib20) [inline strawberryfields/backends/bosonicbackend/backend.py:311]:
```suggestion
                np.array([0.5 * self.circuit.hbar * np.identity(2)]),
            )
```

File: strawberryfields/backends/bosonicbackend/backend.py [hunk 13]
```diff
@@ -187,8 +187,8 @@ def init_circuit(self, prog, **kwargs):
         self.circuit = BosonicModes()
         init_weights, init_means, init_covs = [[0] * nmodes for i in range(3)]
 
-        vac_means = np.zeros(2, dtype=complex).tolist()
-        vac_covs = np.identity(2, dtype=complex).tolist()
+        vac_means = np.zeros((1, 2), dtype=complex)  # .tolist()
```
Comment (nquesada) [inline strawberryfields/backends/bosonicbackend/backend.py:190]:
```suggestion
        vac_means = np.zeros((1, 2), dtype=complex) 
```

**Review by nquesada — APPROVED** [review]

**Pull request merged**